implementations/rust/ockam/ockam_core/src/routing/address.rs

```diff
@@ -78,7 +78,46 @@ pub struct Address {
     pub tt: u8,
     inner: Vec<u8>,
 }
+/// An error which is returned when address parsing from string fails
+#[derive(Debug)]
+pub struct AddressParseError {
+    kind: AddressParseErrorKind,
+}
+/// Enum to store the cause of address parsing failure
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum AddressParseErrorKind {
+    /// Unable to parse address num in the address string
+    InvalidAddr(std::num::ParseIntError),
+    /// Address string has more than one '#' separator
+    MultipleSep,
+}
 
+impl AddressParseError {
+    /// Create new instance
+    pub fn new(kind: AddressParseErrorKind) -> Self {
+        Self { kind }
+    }
+    /// Gets the cause of address parsing failure.
+    pub fn kind(&self) -> &AddressParseErrorKind {
+        &self.kind
+    }
+}
+impl Display for AddressParseError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self.kind {
+            AddressParseErrorKind::InvalidAddr(e) => {
```

Comment:
(`s/InvalidAddr/InvalidType/g` continues...)
```suggestion
            AddressParseErrorKind::InvalidType(e) => {
```